pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/BasePartitionDedupMetadataManager.java

```diff
@@ -175,15 +189,86 @@ public void removeSegment(IndexSegment segment) {
 
   protected abstract void doRemoveSegment(IndexSegment segment, Iterator<DedupRecordInfo> dedupRecordInfoIterator);
 
+  protected boolean isOutOfMetadataTTL(double dedupTime) {
+    return _metadataTTL > 0 && dedupTime < _largestSeenTime.get() - _metadataTTL;
+  }
+
+  protected double getMaxDedupTime(IndexSegment segment) {
+    return ((Number) segment.getSegmentMetadata().getColumnMetadataMap().get(_dedupTimeColumn)
+        .getMaxValue()).doubleValue();
+  }
+
+  /**
+   * Loads watermark from the file if exists.
+   */
+  protected double loadWatermark() {
+    File watermarkFile = getWatermarkFile();
+    if (watermarkFile.exists()) {
+      try {
+        byte[] bytes = FileUtils.readFileToByteArray(watermarkFile);
+        double watermark = ByteBuffer.wrap(bytes).getDouble();
+        _logger.info("Loaded watermark: {} from file: {}", watermark, watermarkFile);
+        return watermark;
+      } catch (Exception e) {
+        _logger.warn("Caught exception while loading watermark file: {}, skipping", watermarkFile);
+      }
+    }
+    return TTL_WATERMARK_NOT_SET;
+  }
+
+  /**
+   * Persists watermark to the file.
+   */
+  protected void persistWatermark(double watermark) {
+    File watermarkFile = getWatermarkFile();
+    try {
```

Comment:
nit: Would it make sense to have a util method as we are reusing same logic in upsert?
  /**
   * Persists watermark to the file.
   */
  protected void persistWatermark(File watermarkFile, double watermark) {
    try {
      if (watermarkFile.exists()) {
        if (!FileUtils.deleteQuietly(watermarkFile)) {
          _logger.warn("Cannot delete watermark file: {}, skipping", watermarkFile);
          return;
        }
      }
      try (OutputStream outputStream = new FileOutputStream(watermarkFile, false);
          DataOutputStream dataOutputStream = new DataOutputStream(outputStream)) {
        dataOutputStream.writeDouble(watermark);
      }
      _logger.info("Persisted watermark: {} to file: {}", watermark, watermarkFile);
    } catch (Exception e) {
      _logger.warn("Caught exception while persisting watermark file: {}, skipping", watermarkFile);
    }
  }